app/controllers/admin/contacts_controller.rb

```diff
@@ -36,17 +36,38 @@ def create
     end
   end
 
+  def delete; end
+
   def destroy
-    if Admin::DestroyContact.new(@contact).call
-      flash.notice = "Contact successfully deleted"
+    redirect_path = remove_govuk_domain(params[:redirect_url])
+    if valid_redirect_url?(redirect_path)
+      if Admin::DestroyAndRedirectContact.new(@contact, redirect_path).destroy_and_redirect
+        flash.notice = "Contact successfully deleted"
+        redirect_to admin_contacts_path
+      else
+        flash.alert = @contact.errors.full_messages.to_sentence
+        redirect_to delete_admin_contact_path
+      end
     else
-      flash.alert = @contact.errors.full_messages.to_sentence
+      flash.alert = "Invalid redirect URL. Redirects must be internal links, e.g. /world"
+      redirect_to delete_admin_contact_path
     end
-    redirect_to admin_contacts_path
   end
 
 private
 
+  def remove_govuk_domain(url)
+    url
+      .sub("https://www.gov.uk", "")
+      .sub("http://www.gov.uk", "")
+      .sub("https://gov.uk", "")
+      .sub("http://gov.uk", "")
+  end
```

Comment:
I wonder if this should use something more generic to return only the path, then it would work in integration and staging too?  E.g.
```suggestion
  def remove_govuk_domain(url)
    URI.parse(url).path
  end
```

As Publishing API [only permits certain domains](https://github.com/alphagov/publishing-api/blob/a0c59b9c7b8ec48fa4256274abed626567aea87a/app/validators/routes_and_redirects_validator.rb#L168-L170), so maybe `valid_redirect_url?` could be called before `remove_govuk_domain` to check that either: a) the URL contains the correct domain for the environment (e.g. obtained from Plek), or b) it is a relative path (rather than trying to trim off pre-defined domains then checking it is a relative path).

Reply:
Thanks for the review, Bruce. It seems reasonable that it should handle integration & staging URLs - I've tacked on a commit. I also like the safety of the `URI.parse` method, rather than the previous "check if it contains any spaces" validation.

I did still need to hard-code a list of allowed hosts, as otherwise `https://food.gov.uk/foo` would be a valid redirect URL, and blindly converted to `/foo`, which would end up being `https://www.gov.uk/foo`, which is not what the publisher intended.

Reply:
Thanks for making these changes.  I'm not a huge fan of having the hardcoded paths in the application though.

Maybe we could do something like this?
```
  def redirect_path(url)
    uri = URI.parse(url)
    return uri.path if uri.host.nil? || uri.host == URI.parse(Plek.new.website_root).host
  rescue URI::InvalidURIError
    false
  end
```

In this case, we'll only match the hosts for the same environment and also stop matching other subdomains of `gov.uk`.

Reply:
Nice idea. I've rewritten the commit. It got slightly hairy because I want to account for if the publisher uses a www-less URL.